src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardSyncerTest.java

```diff
@@ -1857,6 +1862,216 @@ private void testAssertShardCoveredOrAbsentTestIncompleteSplit(HashKeyRange pare
+    }
+
+    private void testCheckAndCreateLeaseForShardsIfMissing(List<Shard> shards,InitialPositionInStreamExtended initialPositionInStreamExtended,
+                                                           Set<String> expectedLeaseKeys) throws Exception {
+        testCheckAndCreateLeaseForShardsIfMissing(shards, initialPositionInStreamExtended, expectedLeaseKeys, Collections.emptyList());
+    }
+
+    private void testCheckAndCreateLeaseForShardsIfMissing(List<Shard> shards,InitialPositionInStreamExtended initialPositionInStreamExtended,
+                                                           Set<String> expectedLeaseKeys, List<KinesisClientLease> existingLeases) throws Exception {
+        final File dataFile = KinesisLocalFileDataCreator.generateTempDataFile(shards, 0, "fileName");
+        dataFile.deleteOnExit();
+        final IKinesisProxy kinesisProxy = spy(new KinesisLocalFileProxy(dataFile.getAbsolutePath()));
+        when(kinesisProxy.getShardList()).thenReturn(shards);
+        when(kinesisProxy.getShardListWithFilter(any())).thenReturn(getFilteredShards(shards, initialPositionInStreamExtended));
+
+        // Populate existing leases
+        for (KinesisClientLease lease : existingLeases) {
+            leaseManager.createLeaseIfNotExists(lease);
+        }
+
+        List<KinesisClientLease> oldLeases = leaseManager.listLeases();
+        shardSyncer.checkAndCreateLeasesForNewShards(kinesisProxy, leaseManager, initialPositionInStreamExtended,
+                false, false);
+        List<KinesisClientLease> newLeases = leaseManager.listLeases();
+        newLeases.removeAll(oldLeases);
+
+        final Set<String> newLeaseKeys = newLeases.stream().map(Lease::getLeaseKey).collect(Collectors.toSet());
+        final Set<ExtendedSequenceNumber> newSequenceNumbers = newLeases.stream().map(KinesisClientLease::getCheckpoint).collect(Collectors.toSet());
+        final Set<ExtendedSequenceNumber> expectedSequenceNumbers = new HashSet<>(Collections
+                .singletonList(new ExtendedSequenceNumber(initialPositionInStreamExtended.getInitialPositionInStream().name())));
+
+        assertThat(newLeases.size(), equalTo(expectedLeaseKeys.size()));
+        assertThat(newLeaseKeys, equalTo(expectedLeaseKeys));
+        assertThat(newSequenceNumbers, equalTo(expectedSequenceNumbers));
+
+        dataFile.delete();
+    }
+
+    /*
+     * Shard structure (each level depicts a stream segment):
+     * 0 1 2 3 4   5- shards till epoch 102
+     * \ / \ / |   |
+     *  6   7  4   5- shards from epoch 103 - 205
+     *   \ /   |  /\
+     *    8    4 9  10 - shards from epoch 206 (open - no ending sequenceNumber)
+     * Missing leases: (0, 6, 8)
+     * Initial position: AT_TIMESTAMP(1000)
+     * Leases to create: (0)
+     */
+    @Test
+    public void testCheckAndCreateLeasesForNewShardsAtTimestampWithPartialLeaseTable1() throws Exception {
```

Comment:
Looks like we're missing the TRIM case for this block of tests

Reply:
Yeah looks like I missed that case, added.